### PR TITLE
11.0 mig account_cutoff_prepaid

### PR DESCRIPTION
**Pull request by ghost**
Depends on:

- [x] account_cutoff_base - #61 
- [x] account_invoice_start_end_dates - #63

### CODE REVIEW

File: account_cutoff_prepaid/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# © 2013-2016 Akretion (Alexis de Lattre <alexis.delattre@akretion.com>)
```
Comment (simahawk) [inline account_cutoff_prepaid/__manifest__.py]: (c) -> Copyright

File: account_cutoff_prepaid/README.rst [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+invoice for my next fiscal year ?*, you will be interested by the
+*forecast* feature. For that, on the Prepaid Revenue form, click on
+the *Forecast* option and you will see 2 new fields: *Start Date* and
+*End Date*. Enter the start date and the end date of your next fiscal
+year and click on the button *Re-Generate lines*: you will see all the
+revenue that you already have in your source journals for that period.
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/89/10.0
```
Comment (simahawk) [inline account_cutoff_prepaid/README.rst]: /11.0

File: account_cutoff_prepaid/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    'name': 'Account Cut-off Prepaid',
+    'version': '11.0.1.0.0',
+    'category': 'Accounting & Finance',
+    'license': 'AGPL-3',
+    'summary': 'Prepaid Expense, Prepaid Revenue',
+    'author': 'Akretion,Odoo Community Association (OCA)',
+    'website': 'http://www.akretion.com',
```
Comment (simahawk) [inline account_cutoff_prepaid/__manifest__.py]: https://github.com/OCA/account-closing

File: account_cutoff_prepaid/models/account_cutoff.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+from odoo.exceptions import UserError, ValidationError
+
+
+class AccountCutoff(models.Model):
+    _inherit = 'account.cutoff'
+
+    @api.model
+    def _get_default_source_journals(self):
+        res = []
+        cutoff_type = self._context.get('type')
```
Comment (simahawk) [inline account_cutoff_prepaid/models/account_cutoff.py]: self.env.context

File: account_cutoff_prepaid/models/account_cutoff.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+            src_journals = self.env['account.journal'].search(
+                [('type', '=', mapping[cutoff_type])])
+            if src_journals:
+                res = src_journals.ids
+        return res
+
+    source_journal_ids = fields.Many2many(
+        'account.journal', column1='cutoff_id', column2='journal_id',
+        string='Source Journals', readonly=True,
+        default=_get_default_source_journals,
```
Comment (simahawk) [inline account_cutoff_prepaid/models/account_cutoff.py]: proxy this w/ a lambda

File: account_cutoff_prepaid/models/account_cutoff.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+        "the prepaid revenue/expense between 2 dates in the future.")
+    start_date = fields.Date(string='Start Date')
+    end_date = fields.Date(string='End Date')
+
+    _sql_constraints = [(
+        'date_type_forecast_company_uniq',
+        'unique('
+        'cutoff_date, company_id, type, forecast, start_date, end_date)',
+        'A cut-off of the same type already exists with the same date(s) !'
+        )]
```
Comment (simahawk) [inline account_cutoff_prepaid/models/account_cutoff.py]: indent back

File: account_cutoff_prepaid/models/account_cutoff.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    @api.constrains('start_date', 'end_date', 'forecast')
+    def _check_start_end_dates(self):
+        for prepaid in self:
+            if prepaid.forecast and prepaid.start_date and prepaid.end_date \
+                    and prepaid.start_date > prepaid.end_date:
+                raise ValidationError(_(
+                    'The start date is after the end date!'))
+
+    @api.onchange('forecast')
+    def forecast_onchange(self):
```
Comment (simahawk) [inline account_cutoff_prepaid/models/account_cutoff.py]: rename to `onchange_forecast`

File: account_cutoff_prepaid/models/account_cutoff.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+            'end_date': end_date_dt,
+            'account_id': aml.account_id.id,
+            'cutoff_account_id': cutoff_account_id,
+            'analytic_account_id': aml.analytic_account_id.id or False,
+            'total_days': total_days,
+            'prepaid_days': prepaid_days,
+            'amount': aml.credit - aml.debit,
+            'currency_id': self.company_currency_id.id,
+            'cutoff_amount': cutoff_amount,
+            }
```
Comment (simahawk) [inline account_cutoff_prepaid/models/account_cutoff.py]: parenthesis one tab back (same for all those occurrences where the closing one does not match visually the opening one)

File: account_cutoff_prepaid/models/account_cutoff.py [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+        # Create mapping dict
+        mapping = mapping_obj._get_mapping_dict(self.company_id.id, self.type)
+
+        # Loop on selected account move lines to create the cutoff lines
+        for aml in amls:
+            line_obj.create(self._prepare_prepaid_lines(aml, mapping))
+        return True
+
+    @api.model
+    def _inherit_default_cutoff_account_id(self):
```
Comment (simahawk) [inline account_cutoff_prepaid/models/account_cutoff.py]: if you proxy the method in the field (as stated in the base PR) you have to remove `_inherit_` here

File: account_cutoff_prepaid/models/account_cutoff.py [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+        account_id = super()._inherit_default_cutoff_account_id()
+        cutoff_type = self._context.get('type')
```
Comment (simahawk) [inline account_cutoff_prepaid/models/account_cutoff.py]: self.env.context

File: account_cutoff_prepaid/tests/test_account_cutoff_prepaid.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        cutoff = self.cutoff_model.create({
+            'cutoff_date': self._date(date),
+            'type': 'prepaid_revenue',
+            'cutoff_journal_id': self.cutoff_journal.id,
+            'cutoff_account_id': self.account_cutoff.id,
+            'source_journal_ids': [(6, 0, [self.purchase_journal.id])],
+        })
+        return cutoff
+
+    def test_0(self):
```
Comment (simahawk) [inline account_cutoff_prepaid/tests/test_account_cutoff_prepaid.py]: give these methods meaningful name :)

**Review by simahawk — CHANGES_REQUESTED** [review]
pls squash translations commits all in one and include base PRs into oca_dependencies.txt to fix the build

File: account_cutoff_prepaid/models/account_cutoff.py (hunk 4 above)
Comment (mpanarin) [inline account_cutoff_prepaid/models/account_cutoff.py]: should also be noted to not forget change `type` here if it will be renamed in base module

**Review by mpanarin — CHANGES_REQUESTED** [review]

**Review by leemannd — APPROVED** [review]
LGTM!
Code review only.

**Comment by simahawk** [comment]
@oleksandrpaziuk  can you update this?

**Comment by fclementic2c** [comment]
I have a error when I click on any cut-off menus : see printscreen
screenshot-3315644-60-f88c0a runbot1 odoo-community org-2018-01-24-15-38-17-562

**Comment by fclementic2c** [comment]
if empty -> take all journals by default


screenshot-3315697-60-89cda9 runbot1 odoo-community org-2018-01-24-16-47-32-525 1

**Comment by fclementic2c** [comment]
:+1: 
ready to be merged  for me

**Comment by fclementic2c** [comment]
@simahawk It is all goo for me. 
I need another approval please

File: account_cutoff_prepaid/models/account_cutoff.py [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+
+        # Loop on selected account move lines to create the cutoff lines
+        for aml in amls:
+            line_obj.create(self._prepare_prepaid_lines(aml, mapping))
+        return True
+
+    @api.model
+    def _default_cutoff_account_id(self):
+        account_id = super()._default_cutoff_account_id()
+        cutoff_type = self.env.context.get('type')
```
Comment (simahawk) [inline account_cutoff_prepaid/models/account_cutoff.py]: @mpanarin hasn't this been renamed to `cutoff_type` in the base module? Pls, adjust any occurrence accordingly (pay attention to not mess up w/ the account `type` w/ a find/replace :wink: )

**Review by simahawk — CHANGES_REQUESTED** [review]
@mpanarin see my comment about the ctx var. Also, as your testing it, would you mind replacing the screenshots? This one is not a blocker: you can submit another PR later... but do not forget about them ;)

**Review by simahawk — APPROVED** [review]

**Review by mpanarin — APPROVED** [review]
LGTM )

**Comment by simahawk** [comment]
@fclementic2c all good :wink:

**Pull request merged**